File: js/t2/tests/model.fuzz.test.js
import { describe, test } from 'node:test';
import assert from 'node:assert/strict';
import { T } from '../model.js';
import * as TI from './testInstance.js';
import { oracle2 } from './oracle.js';
import { assertSnapshotsEqual } from '../../t1/tests/oracle.js';

const eng = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
              TI.PW, TI.FY, TI.FX);

// model.js's snapshot() wraps mg as a read-only { height, width, cell(y,x) }
// accessor (D1, t1/model.js); the oracle's own mg stays a plain nested array
// (it faithfully models T1.v's genuinely bool-valued Grid.L, untouched).
// assertSnapshotsEqual (t1/tests/oracle.js) compares the two representations
// correctly regardless of which shape either side is in.

function randOf(arr) { return arr[Math.floor(Math.random() * arr.length)]; }

function randomTrace(steps) {
  const p0 = randOf(TI.Piece);
  const m = new eng.Machine(p0);
  let prevTotal = 0;
  let prevLevel = 1;
  let prevScore = 0;

  for (let i = 0; i < steps; i++) {
    const kind = Math.floor(Math.random() * 4);
    assert.doesNotThrow(() => {
      if (kind === 0) m.movePiece(randOf([-1, 0]), randOf([-1, 0, 1]));
      else if (kind === 1) m.rotatePiece(Math.random() < 0.5);
      else if (kind === 2) m.fixPiece(randOf(TI.Piece));
      else m.fallStep(randOf(TI.Piece));
    });

    // T1 invariants transfer to the embedded s1 (§2 of proofs.md)
    assert.strictEqual(eng.T1Eng.typeOK(m.s1), true);

    // safe-integer + LevelCorrect + non-negativity
    for (const f of [m.score, m.level, m.combo, m.totalClearedLines])
      assert.ok(Number.isSafeInteger(f));
    assert.strictEqual(m.level, 1 + Math.floor(m.totalClearedLines / 10));
    assert.ok(m.combo >= 0);
    assert.ok(m.score >= 0);
    assert.ok(m.totalClearedLines >= 0);

    // monotonicity
    assert.ok(m.score >= prevScore);
    assert.ok(m.level >= prevLevel);
    assert.ok(m.totalClearedLines >= prevTotal);

    // a clear this step (detected via totalClearedLines delta, not s1.clearedLines,
    // which rides through move/rotate unchanged) strictly raises score, unless
    // score is already saturated at MAX_SAFE_INTEGER.
    if (m.totalClearedLines > prevTotal && prevScore < Number.MAX_SAFE_INTEGER)
      assert.ok(m.score > prevScore);

    prevTotal = m.totalClearedLines;
    prevLevel = m.level;
    prevScore = m.score;
  }
}

describe('T2 model.js fuzz', () => {
  test('long random traces over the test instance: no throw + invariants + monotonicity', () => {
    for (let trial = 0; trial < 20; trial++) randomTrace(1500);
  });

  test('differential oracle agreement over random traces', () => {
    const params = oracle2.makeParams(TI);
    for (let trial = 0; trial < 20; trial++) {
      const p0 = randOf(TI.Piece);
      const m = new eng.Machine(p0);
      let rs = oracle2.initState(p0, params);
      assertSnapshotsEqual(eng.snapshot(m), oracle2.snapshotOf(rs));

      for (let step = 0; step < 150; step++) {
        const kind = Math.floor(Math.random() * 4);
        if (kind === 0) {
          const dy = randOf([-1, 0]), dx = randOf([-1, 0, 1]);
          m.movePiece(dy, dx);
          const r = oracle2.movePiece(dy, dx, rs, params);
          if (r) rs = r;
        } else if (kind === 1) {
          const cw = Math.random() < 0.5;
          m.rotatePiece(cw);
          const r = oracle2.rotatePiece(cw, rs, params);
          if (r) rs = r;
        } else if (kind === 2) {
          const pNew = randOf(TI.Piece);
          m.fixPiece(pNew);
          const r = oracle2.fixPiece(pNew, rs, params);
          if (r) rs = r;
        } else {
          const pNew = randOf(TI.Piece);
          m.fallStep(pNew);
          const r = oracle2.fallStep(pNew, rs, params);
          if (r) rs = r;
        }
        assertSnapshotsEqual(eng.snapshot(m), oracle2.snapshotOf(rs));
      }
    }
  });

  test('adversarial checkAxioms (delegated to T1): malformed params trip an assertion', () => {
    const originalAssert = console.assert;
    let tripped = false;
    console.assert = (cond) => { if (!cond) tripped = true; };
    try {
      tripped = false;
      const raggedGrid = [[false, false], [false]];
      const badEng = T(TI.Piece, raggedGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
                        TI.PW, TI.FY, TI.FX);
      badEng.checkAxioms();
      assert.strictEqual(tripped, true);
    } finally {
      console.assert = originalAssert;
    }
  });
});

File: rs/t2/tests/model_fuzz_test.rs
//! spec: tests/model_fuzz_test.rs — §9, `implementation.md`.

#[path = "test_instance.rs"]
mod fixture;

use fixture::{Prng, TestInstance};
use t1::model::{self as t1_model, Params};
use t2::instance::Tetris;
use t2::model::Machine;

/// Runs `steps` random commands against a fresh `Machine<P>`, asserting no
/// panic and, after every step, `type_ok` on `s1`, `LevelCorrect`, score/
/// level/`totalClearedLines` monotonicity, `ScoreRisesOnClear`, and
/// `s1.gameover` monotonicity.
///
/// §9 — clear detection uses the `totalClearedLines` delta. §9 (oracle
/// scope) / §0.1 — T1's grid-shaped `Correct` conjuncts aren't re-checked
/// here: `t1/tests/model_fuzz_test.rs` covers those on the same engine, and
/// `t2::model` never mutates grid content.
fn fuzz<P: Params>(seed: u64, steps: usize) {
    let mut rng = Prng::new(seed);
    let pieces = P::piece_all();
    let start = rng.choose(pieces);
    let mut m = Machine::<P>::new(start, || rng.choose(pieces));
    let mut prev_score = m.score;
    let mut prev_level = m.level;
    let mut prev_total = m.total_cleared_lines;
    let mut prev_gameover = m.s1.gameover;

    for step in 0..steps {
        match rng.next_below(7) {
            0 => {
                m.move_piece(0, -1);
            }
            1 => {
                m.move_piece(0, 1);
            }
            2 => {
                m.move_piece(-1, 0);
            }
            3 => {
                m.rotate_piece(true);
            }
            4 => {
                m.rotate_piece(false);
            }
            5 => {
                m.fix_piece(rng.choose(pieces));
            }
            _ => {
                m.fall_step(rng.choose(pieces));
            }
        }

        assert!(
            t1_model::type_ok(&m.s1),
            "type_ok failed at step {step}: {:?}",
            m.s1.mg
        );
        assert_eq!(
            m.level,
            1 + m.total_cleared_lines / 10,
            "LevelCorrect failed at step {step}"
        );
        assert!(m.score >= prev_score, "score decreased at step {step}");
        assert!(m.level >= prev_level, "level decreased at step {step}");
        assert!(
            m.total_cleared_lines >= prev_total,
            "totalClearedLines decreased at step {step}"
        );
        if m.total_cleared_lines > prev_total {
            assert!(
                m.score > prev_score,
                "score did not rise on a clear at step {step}"
            );
        }
        assert!(
            !(prev_gameover && !m.s1.gameover),
            "gameover flipped false at step {step}"
        );

        prev_score = m.score;
        prev_level = m.level;
        prev_total = m.total_cleared_lines;
        prev_gameover = m.s1.gameover;
    }
}

#[test]
fn fuzz_test_instance_long_trace() {
    fuzz::<TestInstance>(0x1234_5678_9abc_def0, 20_000);
}

#[test]
fn fuzz_test_instance_many_seeds() {
    for seed in 1u64..=20 {
        fuzz::<TestInstance>(seed, 2_000);
    }
}

#[test]
fn fuzz_real_instance_long_trace() {
    fuzz::<Tetris>(0xdead_beef_cafe_f00d, 5_000);
}
